### ISSUE TITLE
.ProcessName should use QueryFullProcessImageNameW or GetProcessImageFileNameW

### ISSUE THREAD

**Issue by jaykrell** [dotnet/runtime#55685]
On Windows, .ProcessName should use QueryFullProcessImageNameW or GetProcessImageFileNameW.

Presently it used NtQuerySystemInformation(SystemProcessInformation).

The problem with NtQuerySystemInformation(SystemProcessInformation) is:
 - It is tremendously slower, returning _all_ information about _all_ processes,
   instead of _just_ the name, for _one_ process. This is pretty obviously terrible, unless there are very few processes, and _perhaps_ if you only have the process id and not a handle already (two syscalls, open and query). (Too bad Windows doesn't allow query by id to cut the cost.) Enum also has an advantage, sometimes, like, if you actually want more information than just the name. But generally it seems silly.

 - It does not work with terminated processes, which NtQuerySystemInformation(SystemProcessInformation) omit.
   For example, you cannot have a generic:
    WriteLine("The process {0} ended", process.ProcessName);

  unless you call .ProcessName prior to termination to populate the CLR's cache. Such order/cache-sensitivity is also fragile but maybe ok. And that is not even possible: .NET does not expose creating processes suspended, so you will always race with a fast process that terminates quickly. So `.ProcessName` is all but unusable. We are removing our uses pretty thoroughly as a result. Though it is ok on current process, which will never be terminated.

  We have actual code like this, causing us much grief, because..complicated reasons (we actually run mostly on an alternate implementation of NT, which was enumerating terminated processes, which mostly masked these problems, but not entirely; the problem is worse on native NT).
  We are going to reduce our uses of process.ProcessName as a result.

**Comment by danmoseley** [comment]
Seems reasonable, if it works in all the same places. Do you want to offer a PR, @jaykrell ?

**Comment by ryank425** [comment]
I wanted to implement the feature but I do not have an industry experience with calling C++ Win32 codes from C#, so I will just leave where things are for people who will implement this in the future if this is indeed the way to go to fetch ProcessName.

ProcessName exists in https://github.com/dotnet/runtime/blob/a5842801e0c3fef66532da584528905c88b88f5b/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs
```

public string ProcessName
{
    get
    {
        EnsureState(State.HaveProcessInfo);
        return _processInfo!.ProcessName;
    }
}
```

ProcessName ensures that we have _processInfo not be to null. If this is null, EnsureState will make following calls,
 
`_processInfo = ProcessManager.GetProcessInfo(_processId, _machineName);
`

and for Windows, that goes to this file

https://github.com/dotnet/runtime/blob/a5842801e0c3fef66532da584528905c88b88f5b/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Windows.cs#L61

This is where NtProcessManager will call the problematic call

NtQuerySystemInformation from following code

https://github.com/dotnet/runtime/blob/a5842801e0c3fef66532da584528905c88b88f5b/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs#L255


In short, OP doesn't want to call NtQuerySystemInformation in the first place, just to get ProcessName. 
If you can utilize https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-queryfullprocessimagenamew or 
https://docs.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-getprocessimagefilenamew , you can skip EnsureState and return process name.

I am curious about how the benchmark would look by skipping NtQuerySystemInformation.

**Comment by SteveDunn** [comment]
Hi - I'm happy to work on this one.
